### PR TITLE
Project Zenith design branch

### PR DESCRIPTION
**Pull request by shykes**
This is not even close to finished... It is meant to start discussions on the design and marketing of Project Zenith - an initiative to reinvent the Dagger user experience around the CLI, to make it more approachable.

This draft PR contains the following drafts:
* README
* GraphQL schema
* Examples

### Standalone environments

One important aspect of the Zenith design: you use Dagger/zenith to develop environments, and environments have their own source code repository.

This is different from the current best practice, where environment code is embedded into the repository that will be used as its working directory. From the point of view of the environment's programmer, this mixes data and code. In the Zenith design, the environment's code will be cleanly separate from its data.

Note by @tiborvass (from discord)

> To me, it's very analogous to where Dockerfiles are: some dockerized repos will want to have their Dockerfile, while for other repos that don't or are unlikely to ever have Dockerfiles (think source code of gcc or binutils), there are dockerfile-only repos pointing to the external source code.

> Note that even if you have a dockerfile you can always build the repo with another dockerfile of your choosing.

> I'm not exactly sure how zenith wants to separate in practice the source code and the environment.
> What I think is important is that the average user who just wants to contribute to a project and wants to run the jobs, they should have as little friction as possible. So I hope they don't have to specify both the source and the environment.

Note by Conor at Airbyte:

> agree with [Tibor] , I think that solving for the "platform engineer" usecase is important, but lots of people just want to bundle their build logic and repository together, so it should be painless for them to do that while also unlocking all of this extra great functionality. I feel like that is totally doable under this proposal, but might just take some iteration to get right 

From me addressing the questions:

> We have discussed this point at length in person 🙂 Where we have (roughly) settled:

> A resounding YES to supporting both the "advanced platform team" and "I just want an easy repeatable way to build and test my simple software" use cases

> In theory we could achieve this by supporting both "embedded environment code" and "separate environment code". In practice though: we end up only the embedded pattern, and "separate" falls by the wayside. I know this, because that's the situation we're in today.

> Instead we think we can support both use cases by sticking to always separating the environment code, and still making it very easy to support the "easy" use case. Just because "embed in the app repo" is familiar, doesn't mean it's the only possible pattern, or even the best.

> One way we can avoid having to specify the env + app source each time, is with git remote mapping.

### CODE REVIEW

File: zenith/environment.graphqls [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  workdir: Directory!
+
+  "Direct access to the environment as a data graph. Great for scripting and composition"
+  graph: EnvironmentGraph!
+
+  """
+  Extend this environment with the capabilities of other environments.
+  If no namespace is provided, environments are merged - this may cause naming conflicts.
+  """
+  withExtension(namespace: String, environment: Environment!): Environment!
```
Comment (shykes) [inline zenith/environment.graphqls:54]: I kind of feel like this should happen in a static, declarative config file (dagger.json in the environment source, basically).

- I love that I can programmatically load and execute environments from anywhere - github, universe etc.
- But I also would like an easy way to add, remove and generally manage my extensions *without having to write code*. For example in a GUI. Also dependency management is a feature that users expect to have access to directly - to wrap it in code would probably be surprising and confusing to most.

Comment (sipsma) [reply zenith/environment.graphqls:54]: >  or "shadow code" with an increasingly complex DSL.

Agree we should avoid this at all costs. I'm okay with simple stuff like URIs (`local://`, `git://`, `universe://`, etc.), but we should cut it off at that level of complexity.

> But I also would like an easy way to add, remove and generally manage my extensions without having to write code. For example in a GUI. Also dependency management is a feature that users expect to have access to directly - to wrap it in code would probably be surprising and confusing to most.

I think this sort of approach would be viable:
1. The default is to put it in `dagger.json`, we support only simple URIs as mentioned above.
  * This would either be automatically included in your environment or we could require an explicit `WithExtensionsFromConfig` type of call. Automatic inclusion is probably my preference atm.
2. Very advanced use cases that need on-the-fly customization of the extension can still use the explicit `WithExtension` call.
  * This loses the simplicity of the static config, but it would only be needed for advanced users, so I think it's okay to make that tradeoff.
  * Also, if you ever need to do this, it means you are customizing an environment on the fly, which means you could instead just define a new environment that wraps the existing one with your customizations and then load that from your config instead.

tl;dr default to simple config, explicit `withExtension` is reserved for unblocking advanced use cases and has a clear pattern for migrating to the simple config.

Comment (shykes) [reply zenith/environment.graphqls:54]: another possibility, perhaps too crazy: what if we assumed that with the advent of LLMs, software tools can be programmed to query and modify code as easily as a config file. And therefore we don’t need a “dumbed down” version of that code in a simplified json dsl: just regular code for everybody.

Maybe too radical?

File: zenith/README.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+- **Portable**: any machine that can run standard containers can run a DAG.
+- **Familiar**: keep using the tools you know, packaged in containers.
+- **Scriptable**: automate custom tasks in any language, using a powerful API and growing list of SDKs
+- **Extensible**: easily incorporate the capabilities of other environments into your own
+- **Scalable**: all operations in the DAG are scheduled concurrently, which allows seamless parallelization across cores and if needed, across machines in a cluster (*coming soon*)
+- **Fast**: all operations in the DAG are continuously cached, making all computation magically incremental. As a result, daggerizing an environment typically makes builds and other cpu-intensive tasks much faster out of the box.
+
+## What kind of environments?
+
+In theory, any software environment can be daggerized, as long as its individual components can run in a container. In practice, Dagger is most commonly used for non-production environments, such as:
```
Comment (samalba) [inline zenith/README.md:16]: > Dagger is most commonly used for non-production environments

I don't think this is accurate. Do you mean "outside the application runtime" instead of "non-prod environments"? In practice, a release process or a CI systems are both production environments where Dagger is used.

Comment (d3rp3tt3) [reply zenith/README.md:16]: I think this means that CI is part of the development environment and that it isn't a separate entity that sits in limbo between development and production.

Comment (vito) [reply zenith/README.md:16]: I think the disconnect here is that some folks consider CI/CD "production" if it implements the "path to production" whereas this comment is more saying that you don't literally use Dagger to run your actual production environments (like the running webserver etc).

File: zenith/environment.graphqls (hunk 1 above)
Comment (shykes) [reply zenith/environment.graphqls:54]: If we have static file + code, how do we deal with possible conflicts between them?

File: zenith/README.md (hunk 2 above)
Comment (shykes) [reply zenith/README.md:16]: Most people use "production environment" in a sense that does not include their CI. So I used the terminology in that way. But we can make it work either way, it's not core to Dagger or Zenith. We just need to make sure we use the words in a way that doesn't confuse our audience.

File: zenith/environment.graphqls (hunk 1 above)
Comment (sipsma) [reply zenith/environment.graphqls:54]: > Can we have commands for modifying the dagger.json file?

Yeah we currently have `dagger project` for creating/reading dagger.json, but in the old incarnation of `dagger project` last year it also had commands for adding/updating dependencies.

Will take a closer look at the CLI PR soon, but at a first glance looks great!

> If we have static file + code, how do we deal with possible conflicts between them?

Same way that we deal with conflicts within the static file or conflicts within the code, error out.

I normally don't like two sources of truth, but as long as the static file is the "standard blessed way" and the code approach is reserved for our "Advanced Use Cases FAQ" type of documentation, I think it's totally fine. Express as much as you can in the file, if you are the 1% of users who need more flexibility, write some code too.

File: zenith/environment.graphqls [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    sbom: String!
+}
+
+union ArtifactData {
+    Directory
+    File
+    Container
+}
+
+type Checks {
```
Comment (shykes) [inline zenith/environment.graphqls:97]:
```suggestion
type Check {
```

File: zenith/examples/progrock-env/main.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+		WithShell("biome", Biome).
+		// Add env-specific data to the graph,
+		// for scripting, inspection and composition
+		WithGraph(
+			Biome,
+			Generate,
+			Nixpkgs,
+			NixBase,
+			NixDerivation,
+		).
```
Comment (helderco) [inline zenith/examples/progrock-env/main.go:30]: Same as before, not sure how this is meant to be used.

File: zenith/examples/progrock-env/main.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+			func(ctx dagger.Context) (dagger.Check, error) {
+				return new(goTestSuite), nil
+			},
+		).
+		WithExtension("nix", dagger.Git("github.com/vito/nix-env:v1")).
+		WithWarmup(
+			func(ctx dagger.Context, area string) error {
+				Biome.Sync(ctx)
+			},
+		)
```
Comment (helderco) [inline zenith/examples/progrock-env/main.go:41]: Maybe also a `WithCleanup`?
- https://github.com/dagger/dagger/discussions/5404

File: zenith/environment.graphqls (hunk 1 above)
Comment (helderco) [reply zenith/environment.graphqls:54]: You mean to save in `dagger.json` the environment schema when doing `dagger use github.com/vito/progrock@main `? I was thinking that would just save the URL in that file instead.

I get the performance benefits of having a "pre-loaded" schema though but it opens a can of worms in my mind so I need to catch up a bit more to have a more solid idea.

Comment (helderco) [reply zenith/environment.graphqls:54]: One example for an advanced usage of `WithExtension` in code is to load extensions dynamically based on what's in the project. Like buildpacks.
